testdata/vote.txt

```diff
@@ -0,0 +1,97 @@
+# Tests that a node that is behind on the log will lose the vote.
+
+log-level none
+----
+ok
+
+add-nodes 2 voters=(1,2) index=10
+----
+ok
+
+campaign 1
+----
+ok
+
+stabilize
+----
+ok
+
+log-level debug
+----
+ok
+
+raft-state
+----
+1: StateLeader (Voter) Term:1 Lead:1
+2: StateFollower (Voter) Term:1 Lead:1
+
+# Propose a command on 1
+propose 1 foo
+----
+ok
+
+# 2 is now behind on its log.
+raft-log 2
+----
+1/11 EntryNormal ""
+
+# Let 2 campaign. It should fail, because it's behind on its log.
+campaign 2
+----
+INFO 2 is starting a new election at term 1
+INFO 2 became candidate at term 2
+INFO 2 [logterm: 1, index: 11] sent MsgVote request to 1 at term 2
+
+process-ready 2
+----
+Ready MustSync=true:
+Lead:0 State:StateCandidate
+HardState Term:2 Vote:2 Commit:11
+Messages:
+2->1 MsgVote Term:2 Log:1/11
+INFO 2 received MsgVoteResp from 2 at term 2
+INFO 2 has received 1 MsgVoteResp votes and 0 vote rejections
+
+deliver-msgs 1
+----
+2->1 MsgVote Term:2 Log:1/11
+INFO 1 [term: 1] received a MsgVote message with higher term from 2 [term: 2]
+INFO 1 became follower at term 2
+INFO 1 [logterm: 1, index: 12, vote: 0] rejected MsgVote from 2 [logterm: 1, index: 11] at term 2
+
+stabilize
+----
+> 1 handling Ready
+  Ready MustSync=true:
+  Lead:0 State:StateFollower
+  HardState Term:2 Commit:11
+  Entries:
+  1/12 EntryNormal "foo"
+  Messages:
+  1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "foo"]
+  1->2 MsgVoteResp Term:2 Log:0/0 Rejected (Hint: 0)
+  INFO 1 [term: 2] ignored a MsgAppResp message with lower term from 1 [term: 1]
+> 2 receiving messages
+  1->2 MsgApp Term:1 Log:1/11 Commit:11 Entries:[1/12 EntryNormal "foo"]
+  INFO 2 [term: 2] ignored a MsgApp message with lower term from 1 [term: 1]
+  1->2 MsgVoteResp Term:2 Log:0/0 Rejected (Hint: 0)
+  INFO 2 received MsgVoteResp rejection from 1 at term 2
+  INFO 2 has received 1 MsgVoteResp votes and 1 vote rejections
+  INFO 2 became follower at term 2
+> 2 handling Ready
+  Ready MustSync=false:
+  Lead:0 State:StateFollower
+
+raft-state
+----
+1: StateFollower (Voter) Term:2 Lead:0
+2: StateFollower (Voter) Term:2 Lead:0
+
```

Comment:
To demonstrate the difference in before/after behaviour, I propose to add the following here:

1. `set-randomized-election-timeout = electionTimeout` on node 1. This makes the election timeout on node 1 higher than the random timeout no node 2.
2. Tick nodes 1 and a few times. Node 1 will campaign first, and lose / disrupt the cluster again.

After the fix, the behaviour will change: step (2) will cause the node 2 campaign first and win.

Reply:
There might be +-1 error in this idea, but the general idea should be clear.

Reply:
Roughly understood what you mean, but it seems still not working. Actually there isn't a proper value for `set-randomized-election-timeout`.

If you set it as `2*electionTimeout` (assuming it's 2*5=10) on node 1 here,
- before the fix, then 9 ticks won't trigger the campaign for node 1, but it will trigger campaign for node 2 for sure.
- after the fix, node 2's randomized election timeout is set to [10, 15). 9 ticks won't trigger campaign for either node; 10 ticks will trigger campaign for node 1 for sure, but it may also trigger campaign for node 2 (as its minimum possible randomized election timeout is 10).

What I am trying to demonstrate in the test is: once node 2 lost its vote, it will be de-prioritized on the next campaign, so node 1 will start the vote in the following campaign before node 2.

Reply:
Try `2*electionTimeout-1` then. It's ok if the timeout ends up the same on both nodes. Just make sure that node 2 campaigns first anyway. You can arrange this by invoking `ready` on node 2 first.

Reply:
Even more, you don't have to `set-randomized-election-timeout` in the first place. You can do this:

1. tick both nodes `2*electionTimeout-1` times.
2. `process-ready 2`
3. `deliver-msgs 1`
4. `stabilize`

With this sequence, both nodes will campaign, but node 2 tries it first, regardless of the ticking and random timeouts. After all, it doesn't matter which node initiates campaign first - what matters is whose votes are delivered first.

After the fix, this sequence will flip to node 1 campaigning and winning.

Reply:
> * tick both nodes `2*electionTimeout-1` times.
> * `process-ready 2`
> * `deliver-msgs 1`
> * `stabilize`

Yes, it works & thanks. Updated. I agree it makes the review easier by comparing the test changes.